cmd/tink-worker/cmd/root.go

```diff
@@ -0,0 +1,173 @@
+			workerID, _ := cmd.Flags().GetString("id")
+			maxFileSize, _ := cmd.Flags().GetInt64("max-file-size")
+			timeOut, _ := cmd.Flags().GetDuration("timeout")
+			user, _ := cmd.Flags().GetString("registry-username")
+			pwd, _ := cmd.Flags().GetString("registry-password")
+			registry, _ := cmd.Flags().GetString("docker-registry")
+
+			logger.With("version", version).Info("starting")
+			if setupErr := client.Setup(); setupErr != nil {
+				return setupErr
+			}
+
+			ctx := context.Background()
+			if timeOut > 0 {
+				var cancel context.CancelFunc
+				ctx, cancel = context.WithTimeout(ctx, timeOut)
+				defer cancel()
+			}
+
+			conn, err := tryClientConnection(logger, retryInterval, retries)
+			if err != nil {
+				return err
+			}
+			rClient := pb.NewWorkflowSvcClient(conn)
+
+			regConn := internal.NewRegistryConnDetails(registry, user, pwd, logger)
+			worker := internal.NewWorker(rClient, regConn, logger, registry, retries, retryInterval, maxFileSize)
+
+			err = worker.ProcessWorkflowActions(ctx, workerID)
+			if err != nil {
+				return errors.Wrap(err, "worker Finished with error")
+			}
+			return nil
+		},
+	}
+
+	rootCmd.Flags().Duration("retry-interval", retryIntervalDefault, "Retry interval in seconds")
+
+	rootCmd.Flags().Duration("timeout", time.Duration(defaultTimeoutMinutes*time.Minute), "Max duration to wait for worker to complete")
+
+	rootCmd.Flags().Int("max-retry", retryCountDefault, "Maximum number of retries to attempt")
+
+	rootCmd.Flags().Int64("max-file-size", defaultMaxFileSize, "Maximum file size in bytes")
+
+	// rootCmd.Flags().String("log-level", "info", "Sets the worker log level (panic, fatal, error, warn, info, debug, trace)")
+
+	rootCmd.Flags().StringP("id", "i", "", "Sets the worker id")
+	must(rootCmd.MarkFlagRequired("id"))
+
+	rootCmd.Flags().StringP("docker-registry", "r", "", "Sets the Docker registry")
+	must(rootCmd.MarkFlagRequired("docker-registry"))
+
+	rootCmd.Flags().StringP("registry-username", "u", "", "Sets the registry username")
+	must(rootCmd.MarkFlagRequired("registry-username"))
+
+	rootCmd.Flags().StringP("registry-password", "p", "", "Sets the registry-password")
+	must(rootCmd.MarkFlagRequired("registry-password"))
+
+	return rootCmd
+}
+
+// createViper creates a Viper object configured to read in configuration files
+// (from various paths with content type specific filename extensions) and load
+// environment variables that start with TINK_WORKER.
+func createViper() (*viper.Viper, error) {
+	v := viper.New()
+	v.AutomaticEnv()
+	v.SetConfigName("tink-worker")
+	v.AddConfigPath("/etc/tinkerbell")
+	v.AddConfigPath(".")
+	v.SetEnvPrefix("TINK_WORKER")
+	v.SetEnvKeyReplacer(strings.NewReplacer("-", "_"))
+
+	// If a config file is found, read it in.
+	if err := v.ReadInConfig(); err != nil {
+		if _, ok := err.(viper.ConfigFileNotFoundError); !ok {
+			return nil, err
+		}
+	} else {
+		fmt.Fprintln(os.Stderr, "Using config file:", v.ConfigFileUsed())
```

Comment:
Why do we log using a config file, if non was found?

Reply:
If there was an error reading in the config, it could be that none of the config files exist - this error is ok, and is ignored. If there was any other kind of config loading error then we return that error.

If we successfully loaded a config file, we emit that to stderr.

What are you proposing @mmlb? I could log when no config file was found (though none is needed), and I could log when a config file is used, using the common logging format (this PR started before that was introduced and that may explain why I used Println here).

Reply:
I'm passing the logger in now and logging all three possibilities.

Reply:
:+1: